src/net-questdb-client/Ingress/Senders/HttpSender.cs

```diff
@@ -0,0 +1,533 @@
+    {
+        if (WithinTransaction)
+        {
+            throw new IngressError(ErrorCode.InvalidApiCall,
+                "Cannot start another transaction - only one allowed at a time.");
+        }
+        
+        if (Length > 0)
+        {
+            throw new IngressError(ErrorCode.InvalidApiCall,
+                "Buffer must be clear before you can start a transaction.");
+        }
+
+        _buffer.Transaction(tableName);
+        return this;
+    }
+
+    /// <inheritdoc cref="CommitAsync"/> />
+    public void Commit(CancellationToken ct = default)
+    {
+        if (!WithinTransaction)
+        {
+            throw new IngressError(ErrorCode.InvalidApiCall, "No transaction to commit.");
+        }
+        CommittingTransaction = true;
+        Send(ct);
+
+        CommittingTransaction = false;
+        Debug.Assert(!_buffer.WithinTransaction);
+    }
+
+    /// <inheritdoc />
+    public async Task CommitAsync(CancellationToken ct = default)
+    {
+        if (!WithinTransaction)
+        {
+            throw new IngressError(ErrorCode.InvalidApiCall, "No transaction to commit.");
+        }
+        CommittingTransaction = true;
+        await SendAsync(ct);
+
+        CommittingTransaction = false;
+        Debug.Assert(!_buffer.WithinTransaction);
+    }
+
+    /// <inheritdoc />
+    public void Rollback()
+    {
+        if (!WithinTransaction)
+        {
+            throw new IngressError(ErrorCode.InvalidApiCall, "Cannot rollback - no open transaction.");
+        }
+
+        _buffer.Clear();
+    }
+    
+    /// <inheritdoc cref="SendAsync"/>
+    public void Send(CancellationToken ct = default)
+    {
+        if (WithinTransaction && !CommittingTransaction)
+        {
+            throw new IngressError(ErrorCode.InvalidApiCall, "Please `commit` to complete your transaction.");
+        }
+        
+        if (_buffer.Length == 0)
+        {
+            return;
+        }
+        
+        var (request, cts) = GenerateRequest(ct);
+        try
+        {
+            var response = _client!.Send(request, HttpCompletionOption.ResponseHeadersRead, cts!.Token);
+            FinishOrRetry(
+                response, cts
+            );
+        }
+        catch (Exception ex)
+        {
+            inErrorState = true;
+            if (ex is not IngressError)
+            {
+                throw new IngressError(ErrorCode.ServerFlushError, ex.Message, ex);
+            }
+
+            throw;
+        }
+    }
+        
+    /// <inheritdoc />
+    public async Task SendAsync(CancellationToken ct = default)
+    {
+        if (WithinTransaction && !CommittingTransaction)
+        {
+            throw new IngressError(ErrorCode.InvalidApiCall, "Please `commit` to complete your transaction.");
+        }
+        
+        if (_buffer.Length == 0)
+        {
+            return;
+        }
+        
+        var (request, cts) = GenerateRequest(ct);
```

Comment:
Please dispose `request`, `cts` in this method before method exit rather than passing dispose responsibility to `FinishOrRetryAsync`